mailgun/route_handlers.py

```diff
@@ -30,193 +30,222 @@
+    to_list = address.parse_list(request.POST.get('To'))
+
+    logger.debug(u'Handling recipient %s, from %s, subject %s, message id %s',
+                 recipient, sender, subject, message_id)
+
+    # send errors from a no-reply address so we don't get into bounceback loops
+    no_reply_address = settings.NO_REPLY_ADDRESS
+
+    # short circuit if the mailing list doesn't exist
+    try:
+        ml = MailingList.objects.get_or_create_or_delete_mailing_list_by_address(
+                 recipient.address)
+    except MailingList.DoesNotExist:
+        logger.info(
+            u'Sending mailing list bounce back email to %s for mailing list %s '
+            u'because the mailing list does not exist', sender, recipient)
+        template = get_template('mailgun/email/bounce_back_does_not_exist.html')
+        content = template.render(Context({
+            'sender': sender,
+            'recipient': recipient.full_spec(),
+            'subject': subject,
+            'message_body': body_plain or body_html,
+        }))
+        listserv_client.send_mail(no_reply_address, no_reply_address, sender,
+                                  subject='Undeliverable mail', html=content,
+                                  message_id=message_id)
+        return
+
+    # try to determine the course instance, and from there the school
+    school_id = None
+    ci = CourseInstance.objects.get_primary_course_by_canvas_course_id(ml.canvas_course_id)
+    if ci:
+        school_id = ci.course.school_id
+    else:
+        logger.warning(
+            u'Could not determine the primary course instance for Canvas '
+            u'course id %s, so we cannot prepend a short title to the '
+            u'email subject, or check the super senders.', ml.canvas_course_id)
+
+    # get the list of staff and member addresses.  always include staff addresses
+    # in the members list, so they can send to any list in the course, and receive
+    # all mail for the course.
+    teaching_staff_addresses = ml.teaching_staff_addresses
+    member_addresses = teaching_staff_addresses.union(
+                           [m['address'].lower() for m in ml.members])
+
+    # if we can, grab the list of super senders
+    super_senders = set()
+    if school_id:
+        # use iexact here to be able to match on COLGSAS or colgsas
+        query = SuperSender.objects.filter(school_id__iexact=school_id)
+
+        # lowercase all addresses in the supersenders list
+        super_senders = {addr.lower() for addr
+                             in query.values_list('email', flat=True)}
+
+    # if we want to check email addresses against the sender, we need to parse
+    # out the address from the display name.
+    parsed_sender = address.parse(sender)
+    sender_address = parsed_sender.address.lower()
+    sender_display_name = parsed_sender.display_name
+
+    # any validation that fails will set the bounce template
+    bounce_back_email_template = None
```

Comment:
Yup.  Previously, we only sent via `listserv_client` in the "not a list" case, so this was as DRY as it could get.  With the hotfixes over the weekend, there's room for more refactoring.